### ISSUE TITLE
JoinTable.onKey(ForeignKey) does not work when tables are aliased

### ISSUE THREAD

**Issue by SebastianS90** [jOOQ/jOOQ#12214]
### Expected behavior
Consider this simple hierarchy of state, region and city:
```sql
CREATE TABLE STATE
(
    id INT NOT NULL PRIMARY KEY
);

CREATE TABLE REGION
(
    id INT NOT NULL PRIMARY KEY,
    state_id INT NOT NULL,
    CONSTRAINT fk_region_state FOREIGN KEY (state_id) REFERENCES STATE(id)
);

CREATE TABLE CITY
(
    id INT NOT NULL PRIMARY KEY,
    region_id INT NOT NULL,
    CONSTRAINT fk_city_region FOREIGN KEY (region_id) REFERENCES REGION(id)
);
```

The following code
```java
State s = STATE.as("s");
Region r = REGION.as("r");
City c = CITY.as("c");
System.out.println(DSL
        .select(DSL.asterisk())
        .from(s)
        .innerJoin(r).onKey(FK_REGION_STATE)
        .innerJoin(c).onKey(FK_CITY_REGION)
        .getSQL()
);
```

should produce this query:
```sql
select * from "STATE" "s" join "REGION" "r" on "r"."STATE_ID" = "s"."ID" join "CITY" "c" on "c"."REGION_ID" = "r"."ID"
```

### Actual behavior
The resulting query is
```sql
select * from "STATE" "s" join "REGION" "r" on "r"."STATE_ID" = "s"."ID" join "CITY" "c" on "c"."REGION_ID" = "s"."ID"
```
(note the **s** instead of **r** at the end) and it logs the following information
```
INFO: Ambiguous match found for ID. Both "s"."ID" and "r"."ID" match.
java.sql.SQLWarning
	at org.jooq.impl.FieldsImpl.field0(FieldsImpl.java:274)
	at org.jooq.impl.FieldsImpl.field(FieldsImpl.java:213)
	at org.jooq.impl.AbstractRow.field(AbstractRow.java:238)
	at org.jooq.impl.FieldsTrait.field(FieldsTrait.java:67)
	at org.jooq.impl.JoinTable.onKey(JoinTable.java:749)
	at org.jooq.impl.JoinTable.onKey(JoinTable.java:735)
	at org.jooq.impl.JoinTable.onKey(JoinTable.java:149)
	at org.jooq.impl.SelectQueryImpl.addJoinOnKey(SelectQueryImpl.java:4405)
	at org.jooq.impl.SelectImpl.onKey(SelectImpl.java:2221)
	at org.jooq.impl.SelectImpl.onKey(SelectImpl.java:163)
	at Main.main(Main.java:16)
```
originating from [here](https://github.com/jOOQ/jOOQ/blob/version-3.15.1/jOOQ/src/main/java/org/jooq/impl/FieldsImpl.java#L274).

As far as I can tell it should actually already return in [this block](https://github.com/jOOQ/jOOQ/blob/version-3.15.1/jOOQ/src/main/java/org/jooq/impl/FieldsImpl.java#L228-L234), never reaching the lower part. It all works fine without aliases, so I guess that the `equals` method invoked in that block does not take aliases into account.

### Steps to reproduce the problem

[Demo repo](https://github.com/SebastianS90/jooq-onkey-condition/tree/exact-match-check) (select the `exact-match-check` branch)

### Versions

- jOOQ: 3.15.1
- Java: 11

**Comment by lukaseder** [comment]
Thanks for your report. The main problem here is that you're aliasing your tables, but not your constraints. There will be ambiguities that are hard to resolve in some cases, even though it looks as though in your case it should be simple, because the constraint reference references tables that are unambiguous in the join tree (i.e. there are no self-joins)

A related problem is that there's no way yet to make aliased tables reference alias-referencing constraint definitions, see https://github.com/jOOQ/jOOQ/issues/5050

I think that given there are no self-joins, the situation can be improved. Will investigate

**Comment by lukaseder** [comment]
Note we have an integration test for this that doesn't fail, though it does log the ambiguous columns warning. The failure probably depends on ordering of columns in table definitions and in the join tree.

**Comment by SebastianS90** [comment]
Note that the same issue occurs if you let jOOQ find out the key itself (`.onKey()` without parameter).

If your integration test logs that warning it is probably just pure luck that it took the correct table-column for the join (as you said, depending on the order of duplicate names like `id` in my example). Can I help in any way?

**Comment by lukaseder** [comment]
> Can I help in any way?

No that should be fine, thanks. I'll reproduce it with your demo repo tomorrow, and/or try reproducing it by enumerating all possible orderings of tables / columns / etc...

**Comment by lukaseder** [comment]
Related: https://github.com/jOOQ/jOOQ/issues/14668

**Comment by lukaseder** [comment]
This particular case might be fixed by fixing one of the root causes: https://github.com/jOOQ/jOOQ/issues/14671

There will still be configurations where `onKey(ForeignKey)` reference usage won't work, because we cannot alias the tables in a `ForeignKey` reference to disambiguate things.

**Comment by lukaseder** [comment]
Unfortunately, #14671 still doesn't fix this. There's yet another lookup issue involved here, so I'll fix this separately.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.18.0, 3.17.9 (https://github.com/jOOQ/jOOQ/issues/14678), 3.16.15 (https://github.com/jOOQ/jOOQ/issues/14679)
